src/styled.js

```diff
@@ -1,29 +1,40 @@
 import { css } from './css';
 
-let h;
-const setPragma = pragma => (h = pragma);
+let h, forwardRef, useTheme;
+const setup = (_h, _forwardRef, _useTheme) => {
+    h = _h;
+    forwardRef = _forwardRef;
+    useTheme = _useTheme;
+};
 
 /**
  * Styled function
  * @param {String} tag
  */
 function styled(tag) {
     const _ctx = this || {};
-
     return function() {
         const _args = arguments;
+        function Styled(props, ref) {
+            _ctx.p = {
+                ...(useTheme && { theme: useTheme() }),
+                ...props
+            };
 
-        return function Styled(props) {
-            const _props = (_ctx.p = Object.assign({}, props));
-            const _previousClassName = _props.className;
-
+            const _previousClassName = _ctx.p.className;
             _ctx.o = /\s*go[0-9]+/g.test(_previousClassName);
-            _props.className =
-                css.apply(_ctx, _args) + (_previousClassName ? ' ' + _previousClassName : '');
+            const _props = {
+                ...props,
+                ref,
+                className:
+                    css.apply(_ctx, _args) + (_previousClassName ? ' ' + _previousClassName : '')
+            };
 
             return h(tag, _props);
```

Comment:
```suggestion
            return h(tag, Object.assign({ ref, className: css.apply(_ctx, _args) + (_previousClassName ? ' ' + _previousClassName : '')  }, props));
```

This removes the need for line 26-31